### PR TITLE
Fix for Issue #636

### PR DESCRIPTION
**Pull request by mkeilman**
Main features here: new routing with a folder path parameter, file tree persistence, and recovery of the stored path from the URL

### CODE REVIEW

File: sirepo/package_data/static/js/sirepo.js [hunk 1]
```diff
@@ -1907,13 +1963,35 @@ SIREPO.app.controller('SimulationsController', function (appState, panelState, r
...
         self.activeFolder = item;
         self.activeFolderPath = [];
         while (item) {
             self.activeFolderPath.unshift(item);
             item = item.parent;
         }
         appState.setActiveFolderPath(self.pathName(self.activeFolder));
+        appState.setActiveFolder(self.activeFolder);
+        if( prevPath !== self.pathName(self.activeFolder) ) {
+            var compoundPath = pathToCompoundPath(self.pathName(self.activeFolder));
```
Comment (robnagler) [inline sirepo/package_data/static/js/sirepo.js]: pathToCompoundPath can return '/' so it seems if you want to check if it is root folder first before calling compoundPath.

File: sirepo/package_data/static/js/sirepo.js [hunk 2]
```diff
@@ -2147,4 +2225,47 @@ SIREPO.app.controller('SimulationsController', function (appState, panelState, r
         appState.saveQuietly('simulationFolder');
     });
     loadList();
+
+    // invoked in loadList() callback
+    function checkURLForFolder() {
+
+        var pathNoSectionRE = new RegExp('^\/' + activeSection.getActiveSection());
```
Comment (robnagler) [inline sirepo/package_data/static/js/sirepo.js]: This shouldn't be a regular expression. getActiveSection needs to be escaped if it is, but more importantly, it's literal match below.

File: sirepo/package_data/static/js/sirepo.js [hunk 3]
```diff
@@ -2147,4 +2225,47 @@ SIREPO.app.controller('SimulationsController', function (appState, panelState, r
...
+                appState.setActiveFolderPath(canonicalPath);
+                appState.setActiveFolder(newFolder);
+                self.openItem(newFolder);
+            }
+            else {
+                requestSender.localRedirect('notFound');
+            }
+        }
+        else {
+            if( !appState.getActiveFolder() ) {
```
Comment (robnagler) [inline sirepo/package_data/static/js/sirepo.js]: Seems like this check needs to be first. The root case would eliminate the getActiveFolderPath check (|| '/') above.

File: sirepo/package_data/static/js/sirepo.js [hunk 4]
```diff
@@ -2147,4 +2225,47 @@ SIREPO.app.controller('SimulationsController', function (appState, panelState, r
...
+                self.openItem(rootFolder());
+            }
+        }
+    }
+    function folderForPathInList (folderPath, folderList) {
+        for( var i = 0; i < folderList.length; ++i ) {
+            var item = folderList[i];
+            if( item.isFolder ) {
```
Comment (robnagler) [inline sirepo/package_data/static/js/sirepo.js]: short circuit

File: sirepo/package_data/static/js/sirepo.js [hunk 5]
```diff
@@ -2147,4 +2225,47 @@ SIREPO.app.controller('SimulationsController', function (appState, panelState, r
...
+                var lPath = self.pathName(item);
```
Comment (robnagler) [inline sirepo/package_data/static/js/sirepo.js]: I'm confused about compoundPath vs pathName. Seems like they are similar.

File: sirepo/package_data/static/js/sirepo.js [hunk 6]
```diff
@@ -1907,13 +1963,35 @@ SIREPO.app.controller('SimulationsController', function (appState, panelState, r
...
+                var params = {
+                    ':folderPath?': compoundPath,
+                };
+                requestSender.localRedirect('simulationsFolder', params);
+            }
+        }
+    }
+
+    function pathToCompoundPath(path) {
+        return path.replace(/^\//,'').replace(/\//g, '_');
```
Comment (robnagler) [inline sirepo/package_data/static/js/sirepo.js]: I don't understand this replacement. '_' is a legitimate path character so it can't be used as a separator. These two paths are equal as compound paths: '/a_b' and '/a/b'. However, that's incorrect. You have to use a character that is not a valid path character as a separator. Not sure why compound path is needed at all.

**Comment by moellep** [comment]
I think it may be worth separating out the folder functionality into a separate factory now. The appState factory's API grew by 2 private vars and 6 public methods. Making a new factory, like fileManager, would make it easier to follow the code.

File: sirepo/package_data/static/js/sirepo.js (hunk 6 above)
Comment (mkeilman) [reply sirepo/package_data/static/js/sirepo.js]: Right, I meant to ask what to use.  I would advocate for something that does not get url encoded...how about "!" like in old-school email addresses?  Then just add it to the list of forbidden characters

File: sirepo/package_data/static/js/sirepo.js [hunk 7]
```diff
@@ -132,6 +137,16 @@ SIREPO.app.factory('appState', function(errorService, requestSender, requestQueu
     var savedModelValues = {};
     var activeFolderPath = null;
 
+    var appActiveFolder = null;
```
[the remark this answers is not in the document]
Comment (mkeilman) [reply sirepo/package_data/static/js/sirepo.js]: For my own sanity so I could keep things straight while I was jumping around in the code, I'll revert it

File: sirepo/package_data/static/js/sirepo.js (hunk 6 above)
Comment (robnagler) [reply sirepo/package_data/static/js/sirepo.js]: Let's not add more invalid chars. We already have a good enough list, e.g. could use ':'. 

I wonder about compoundpath, however. why is it necessary?

Comment (mkeilman) [reply sirepo/package_data/static/js/sirepo.js]: I think angular can't find the route if we use "/" -- but I may be misinterpreting your question. BTW I believe ":" will fail for a similar reason (it will be interpreted as a route param) but I'll try it

File: sirepo/package_data/static/js/sirepo.js [hunk 8]
```diff
@@ -1856,35 +1900,47 @@ SIREPO.app.controller('SimulationsController', function (appState, panelState, r
...
-                    self.openItem(showItem.parent);
-                }
-                else {
-                    self.openItem(rootFolder());
-                }
+                checkURLForFolder();
             });
     }
+    // compares simulationID only
+    function isSimInArray(sim, arr) {
```
[the remark this answers is not in the document]
Comment (mkeilman) [reply sirepo/package_data/static/js/sirepo.js]: I guess we don't even need the data objects in their entirety; might as well store only the sim id. indexOfSimInArray was there for deletion purposes (as referenced in your other comment) but I forgot to complete that task, bah. Will clean that up

File: sirepo/package_data/static/js/sirepo.js (hunk 6 above)
Comment (mkeilman) [reply sirepo/package_data/static/js/sirepo.js]: Followup - colon is fine as it gets url encoded. I find the resulting URLs hard to parse visually, but barring user feedback to that effect it doesn't matter I suppose

File: sirepo/package_data/static/js/sirepo.js (hunk 5 above)
Comment (mkeilman) [reply sirepo/package_data/static/js/sirepo.js]: They are identical except for the delimiter. I guess pathName is just used to depict an artificial file system on the UI so we could change its delimiter and eliminate compoundPath - but I'd want to confirm that's OK!

File: sirepo/package_data/static/js/sirepo.js (hunk 4 above)
Comment (mkeilman) [reply sirepo/package_data/static/js/sirepo.js]: `            if( !item.isFolder ) {
                continue;
            }
...
`
?

Comment (robnagler) [reply sirepo/package_data/static/js/sirepo.js]: Yes, but not one line:

```
if (! item.isFolder) {
    continue;
}
```

Note the spacing, btw. We put a space after the keyword and the paren, no spaces between the paren and the next token, space after `!` and the value it is operating on so it is more visible.

File: sirepo/package_data/static/js/sirepo.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (robnagler) [reply sirepo/package_data/static/js/sirepo.js]: Yes, that's true. I think Paul's advice is the best: separate out in to a factory. The rest may fall out of that.

File: sirepo/package_data/static/js/sirepo.js (hunk 6 above)
Comment (robnagler) [reply sirepo/package_data/static/js/sirepo.js]: Given the way angular works, we might be better off using a search (query) parameter:

```text
/simulations?p=/a/b/c
```

This will produce a readable URL. If angular had the concept of path info, we'd be fine, but it doesn't. Using "search" has the added benefit of not requiring two routes as empty searches work just fine.

**Comment by robnagler** [comment]
@moellep I think you should review. I told @mkeilman we shouldn't spend more time on this unless there are bugs.

**Comment by moellep** [comment]
I've merged in the code and added commit 8ffed6d to move the compoundPath code into fileManager. More of the code in simulationsController could also move to fileManager. I needed to do this so the "Simulations" link would return to the original folder when returning from a simulation to the list.
